### PR TITLE
Add new test cases and refactor RowMergerTest.

### PR DESCRIPTION
**Pull request by gamorris**
Adds a RowMatcher API for matching Row objects.  Refactors RowMergerTest
to use the new row matcher API.  Adds some new test cases for row reset
scenarios.

### CODE REVIEW

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java [hunk 1]
```diff
@@ -72,135 +66,132 @@ public void resultsAreReadable() throws IOException {
...
-    Assert.assertEquals(ByteString.copyFromUtf8("c1"),
-        resultFamily1.getColumns(0).getQualifier());
-    Assert.assertEquals(ByteString.copyFromUtf8("c2"),
-        resultFamily1.getColumns(1).getQualifier());
-
-    Family resultFamily2 = familyMap.get("Family2");
-    Assert.assertEquals(0, resultFamily2.getColumnsCount());
-
-    Assert.assertFalse(iterator.hasNext());
+    Assert.assertFalse("Responses not completely consumed by expected rows", iterator.hasNext());
```
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java]: Assert.assertEquals(responses.length, expected_rows.length)

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+      String sep = "";
+      for (Matcher<Cell> m : cells) {
+        description.appendText(sep).appendDescriptionOf(m);
+        sep = ", ";
+      }
+    }
+
+    @Override
+    protected boolean matchesSafely(Column item) {
+      if (!qualifier.matches(item.getQualifier()))
```
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java]: Please add curly braces around if statement bodies.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcherTest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    builder.addFamilies(family);
+
+    Row row = builder.build();
+    // Not enough families expected.
+    assertThat(row, not(matchesRow("row-1").withFamily("some family")));
+    // Correct.
+    assertThat(row, matchesRow("row-1").withFamily("some family").withFamily("other family"));
+    // Too many families expected.
+    assertThat(row, not(matchesRow("row-1")
+        .withFamily("some family").withFamily("other family").withFamily("third family")));
```
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcherTest.java]: It's easier to read if each .withXX() is on its own line.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+   * Adds a column with qualifier matching the given matcher to the current
+   * column family.  A column family <em>must</em> have already been added
+   * to the expected row with {@link #withFamily(String)} prior to calling
+   * {@code withColumn}.
+   *
+   * Example:
+   *   {@code withFamily("family").withColumn(any(ByteString.class))}
+   */
+  public RowMatcher withColumn(Matcher<ByteString> columnName) {
+    lastColumn = new ColumnMatcher(columnName);
```
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java]: hm.  It's interesting to be able to do withColumn(..).withCell(..), but it seems a tad magical.  Perhaps withColumn(name, cellValue)?

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java [hunk 5]
```diff
@@ -72,135 +66,132 @@ public void resultsAreReadable() throws IOException {
...
 
-    Map<String, Family> familyMap = new HashMap<>();
-    for (Family family : resultRow.getFamiliesList()) {
-      familyMap.put(family.getName(), family);
+  private void matchResponses(ReadRowsResponse[] responses, RowMatcher[] expectedRows) {
+    Iterator<ReadRowsResponse> iterator = getIterator(responses);
+    for (int i = 0; i < expectedRows.length; ++i) {
+      Assert.assertTrue("More expected rows than supplied responses", iterator.hasNext());
+      Row result = RowMerger.readNextRow(iterator);
+      Assert.assertThat("Expected row does not match!", result, expectedRows[i]);
```
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java]: Will this return more details about what didn't match?

**Comment by sduskis** [comment]
I added some nits.  That said, I really like the approach.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java (hunk 4 above)
Comment (gamorris) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java]: The challenge is what to do with multiple cells in a column? Or timestamps? or etc...

Using the "willCell" approach, multiple cells are:
 withColumn(name).withCell(cell1).withCell(cell2).withCell(cell3);

Alternatively: withColumn(name, cell1, cell2, cell3)
The question is, if you go that route then why not: withFamily(name, column1, column2)?
Why not: withRow(rowkey, family1, family2)?

That was actually my first version of this API. It winds up looking like:
assertThat(thing, matchesRow(key, family(name, column(columnName, cell1, cell2))));

...which is not awful... but I like withCell() better, even if it does seem a bit magical.

Thoughts?

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java (hunk 5 above)
Comment (gamorris) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java]: Yes; you get an AssertionError like:

java.lang.AssertionError: Expected row does not match!
Expected: row with key matching <<ByteString@379619aa size=5>> and column families: family name: "Family3", family name: "Family1""column qualifier "<<ByteString@cac736f size=2>> with <1> cells matching: an instance of com.google.bigtable.v1.Cell, "column qualifier "<<ByteString@5e265ba4 size=2>> with <1> cells matching: an instance of com.google.bigtable.v1.Cell
     but: was <key: "row-1"
families {
  name: "Family2"
}
families {
  name: "Family1"
  columns {
    qualifier: "c1"
    cells {
      timestamp_micros: 100
      value: "\2324.\223)dF\241\344\312"
    }
  }
  columns {
    qualifier: "c2"
    cells {
      timestamp_micros: 100
      value: "\210\302w\251Tr\271\031>\276+\315\362\344V$\247\000\322\305\275\364pc6\250\026)\276 \232\333%e\233\302\356\270A\206T\376S\226\260\264#q\317h\265\016\370j\244\363L\312\335\002\273r\034\214\'l\342j\2336\230\022\2433\362\017\372Dl\306wu\267Q\310\255\306D\356=\'}\322\341\364p(<\243\213"
    }
  }
}
>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at com.google.cloud.bigtable.grpc.scanner.RowMergerTest.matchResponses(RowMergerTest.java:78)
	at com.google.cloud.bigtable.grpc.scanner.RowMergerTest.multipleChunksAreMerged(RowMergerTest.java:85)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[etc.]


It is not the greatest error in the world. (Especially don't care for the ByteString representation.) But making that better seems of limited value at the moment...

Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMergerTest.java]: nice

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java (hunk 2 above)
Comment (gamorris) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java]: done.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcherTest.java (hunk 3 above)
Comment (gamorris) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcherTest.java]: See if you like it better now.

File: bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java (hunk 4 above)
Comment (sduskis) [inline bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/scanner/RowMatcher.java]: gotcha.  Works for me.

**Comment by sduskis** [comment]
LGTM.  Please squash the commits: https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/CONTRIBUTING.md#squashing-commits

**Pull request merged**